### PR TITLE
Add backtrace support to derive-Error macro

### PR DESCRIPTION
**Pull request by ffuugoo**
Implementation is mostly done, but there are 3 unresolved issues:

- [The way `State` process attributes need some reorganization.][state-attribute-processing-issue] Right now I've worked around it [by sticking conditional on "deriving Error" right into `State` code][state-attribute-processing-issue-workaround].
- I'm not sure what is the proper way to [(unstable-)feature-gate macro-generated code][macro-generated-code-feature-gating].
- I'm not sure what is the proper way to enable/disable tests for unstable features (e.g., backtrace) [[1]][tests-feature-gating-1] [[2]][tests-feature-gating-2].

[state-attribute-processing-issue]: https://github.com/JelteF/derive_more/issues/92#issuecomment-556081021
[state-attribute-processing-issue-workaround]: https://github.com/JelteF/derive_more/pull/110/files#diff-3dbd16223f25e8b5e0bcf099743107e9R384
[macro-generated-code-feature-gating]: https://github.com/JelteF/derive_more/pull/110/files#diff-9a269209bed069ffb36d5565a91c9840R241
[tests-feature-gating-1]: https://github.com/JelteF/derive_more/pull/110/files#diff-42b35eae59c844e338d41162d1653d1aR1
[tests-feature-gating-2]: https://github.com/JelteF/derive_more/pull/110/files#diff-b62cdc7ae8f9b197b5f1b1ec1abae2d6R51

### CODE REVIEW

**Comment by ffuugoo** [comment]
@JelteF @tyranron

File: Cargo.toml [hunk 1]
```diff
@@ -133,7 +133,7 @@ required-features = ["display"]
 
 [[test]]
 name = "error"
-path = "tests/error.rs"
+path = "tests/error_tests.rs"
```
Comment (ffuugoo) [inline Cargo.toml]: Had to change file name, as I've moved tests to separate module and Rust won't allow me to have `mod error` in `error.rs`.

**Comment by JelteF** [comment]
For the last two questions, you can revert part of this commit:
https://github.com/JelteF/derive_more/commit/ca23e47307d71c876871f5b5749b5d6dccdaa285
That does automatic detection of nightly. I think for tests t's needed for sure. For actual library code I don't know if we want that at all, because as long as you don't have a backtrace field or `Backtrace` type you won't generate backtrace code. That way if you try to generate backtrace code with a non-nighthly compiler you will get a compiler error saying you should run nightly. 

For your first issue I think there's two issues:
1. Your example in the original issue should work without the `source` attribute. The `Backtrace` field should be ignored by default as as a `source`.
2. With three fields (e.g `E(Error, i32, Backtrace)`) the issue still stands and there should be a solution. I think your solution is slightly hacky but does the job. I'll think a bit more if I can think of a more elegant solution.

FYI I won't have time to review this until mid december.

**Comment by ffuugoo** [comment]
> as long as you don't have a backtrace field or Backtrace type you won't generate backtrace code

I thought the same at first, but that's not entirely true. Consider this situation:

```rust
#[derive(Error)]
struct MyError {
    source: SomeOtherError,
}
```

Now, consider, that `SomeOtherError` has a `backtrace` method that actually returns some backtrace. If we completely discard generating `backtrace` method if no backtrace-field specified, then we would have to explicitly mark `source` field with `#[error(backtrace)]` attribute, if we want to generate `backtrace` method.

It's not the biggest problem, but ___if there is a way to properly feature-gate generated `backtrace` method___ (and I'm not sure it's possible), then I'd prefer to generate feature-gated method if we have either source-field and/or backtrace-field.

**Comment by ffuugoo** [comment]
> Your example in the original issue should work without the source attribute. The Backtrace field should be ignored by default as as a source.

Seems reasonable. I'll improve this part. But the issue still stands even for two fields case. Consider:

```rust
#[derive(Error)]
struct MyError(#[error(not(backtrace))] SomeOtherError, Backtrace);
```

Specifying `#[error(not(backtrace))]` on the source-field explicitly disables backtrace-chaining (i.e., when we try to retrieve backtrace from source-field and if it returns `None` return backtrace stored in structure/enum-variant itself).

However, in this case, specifying valid attribute on a source-field will mark all other fields as "disabled" and we'll get nothing.

**Comment by ffuugoo** [comment]
> I think your solution is slightly hacky but does the job.

It's absolutely hacky. I just put it there as a temporary solution, as reworking `State` (and other macro) is a big task in itself and I wanted to more-or-less finalize my solution on `Error`.

**Comment by ffuugoo** [comment]
>Your example in the original issue should work without the source attribute.

Fixed.

**Comment by ffuugoo** [comment]
I've experimented a bit with generating feature-gated code and come to conclusion that it's impossible to do properly in our case.

It's possible to generate feature-gated code. And using `rustversion` crate it's even possible (not at all trivial, though) to feature-gate by compiler channel. But in the end it's impossible to feature-gate by ___enabled unstable feature___. So, we can't generate implementation that would automatically become available when `#![feature(backtrace)]` is enabled.

I suppose we should add separate `error-backtrace` (or just `backtrace`) feature and [apply it in implementation instead of `nightly`][implementation-nightly-feature-gate].

And tests should stay as is (i.e., feature-gated by `nightly` and `#![cfg_attr(feature = "nightly", feature(backtrace)]`).

And then we can also optionally revert `rustc_version` check in `build.rs` and force-enable `nightly` feature from there so that full test-suite is ___guaranteed___ to run on nightly.

[implementation-nightly-feature-gate]: https://github.com/JelteF/derive_more/pull/110/files#diff-9a269209bed069ffb36d5565a91c9840R241

**Comment by JelteF** [comment]
Hmm, this indeed sounds like a shitty situation. One thing that would solve
it though afaict, is if the default implementation of backtrace would
forward to self.source() instead of always returning None. I think that
would be a very reasonable thing to do. Could you create a PR for this on
rust itself? This should still be possible to change there, since it is
still an unstable API.

**Comment by ffuugoo** [comment]
@JelteF I've seen your questions on RFC 2504. What's your current opinion on all of this? I'd say that dropping the whole backtrace-chaining thing for now is the most safe thing to do.

**Comment by ffuugoo** [comment]
Also, as an idea for either derive-Error or a separate crate, what if we implement formatting for `std::error::Error` types in a way [formatting is being implemented for `std::backtrace::Backtrace`](https://github.com/rust-lang/rust/issues/65280)? I.e., use some format arguments to print causes/sources.

**Comment by ffuugoo** [comment]
> I'll re-review the PR and probably try to chart all implicit/explicit source/backtrace permutations

I've refreshed my context. Seems like tests I've made are good enough specification of expected behaviour.

---

Considering my proposal to "not generate backtrace method implementation _implicitly_ when there is only source-field, but no backtrace-field": 

We have to change implementation (and tests) in such way, so:

- all `*_no_backtrace_source_with_backtrace_explicitly_disabled` tests are removed
  (as it would be default behaviour and explicit disabling won't make any difference)
- all `*_no_backtrace_source_with_backtrace` test cases would return `None` as backtrace
  (as no backtrace-method implementation would be generated and default implementation returns
  `None`)
- a set of `*_no_backtrace_source_with_backtrace_explicitly_enabled` test cases would be added
  and all of them would return `Some(Backtrace)` as backtrace

We also have to make sure that we do actually chain backtrace calls in cases when:

- there is a source attribute without explicit `#[error(not(backtrace))]` attribute specified
- there is a backtrace-field (inferred or explicitly specified)

But current test suite should catch such errors.

As far as I can see, this should not be too difficult to implement and it also won't introduce any "impossible-to-specify" or "ambiguous" attribute combinations.

**Comment by JelteF** [comment]
Thanks for picking this up again. Based on the discussion I've had on RFC 2504 I actually think that it would be better if we don't implement any backtrace forwarding. This is an antipattern that I don't think this library should try to make easy.

So i think we should only support a backtrace field, inferred or explicitely specified in cases where it cannot be inferred.

**Comment by ffuugoo** [comment]
> This is an antipattern that I don't think this library should try to make easy.

I have mixed ideas in this regard.

It _may_ be an antipattern, but:

- so far the only compelling case I've seen when such behaviour may be harmful is when we have an _'error with backtrace'_ have an _'error with backtrace __from another thread__'_ as its source

  - though I'd assume it's less common case

  - backtrace forwarding can be disabled with attributes, so derive-Error still can be used for such errors with very moderate discomfort

- backtrace-chaining is, probably, the most common case of all

- `std::backtrace::Backtrace` is still unstable and it's unclear for me if community would settle for _'backtrace chaining as usual default'_ or _'`std::error::Error::backtrace` never chains, chaining implemented with extension-traits'_

  - and, if backtrace-chaining happens to be widespread, being a purist against a pattern used by most of community seems impractical

- dtolnay decided to implement backtrace-chaining in `thiserror`, so, at least one respected Rust developer thinks it's not that bad

  - 'argument from authority' is a logical fallacy, but still...

- we can always change it later, if it would prove to be a faulty decision

  - though, I recognize, that breaking changes for a crate with 1.5 millions downloads may be a potentially questionable decision

@tyranron may also have some valid arguments.

**Comment by JelteF** [comment]
I think the responses of @mitsuhiko explain why it's an anti pattern pretty
clearly here: https://github.com/rust-lang/rust/issues/53487

But for completeness I'll reword it here too. The main reason why it's an
anti pattern, is that it causes problems if you want to show the whole
error chain (which is usually what you want). In that case you suddenly
cannot show all the stacktraces in the chain, because there will be
duplicate ones in the chain.

Users don't really care about easy backtrace forwarding. They care about
easy printing of an error chain with backtraces. As long as you don't
implement backtrace forwarding, this function is very easy to implement. If
you don't it becomes much harder/impossible. I'd like this library to help
bring the ecosystem in a good state, where we have error chains that are
easy to display nicely. Not the other way around.

So especially with regards to backwards compatibility: I think we should
remove backtrace forwarding. Removing it later would be a breaking change.
Adding it later should be no problem, if it turns out that we do need
forwarding.

**Comment by tyranron** [comment]
@JelteF I understand why backtrace forwarding is an anti-pattern when is applied blindly and by default. I agree. Let's just not apply any backtrace forwarding by default. But I cannot understand why we should remove the ability when library user _really requires backtrace forwarding_ in his case? Let's make it just opt-in: apply the forward behavior _only if_ `#[error(backtrace(forward))]` attribute is applied to the field explicitly. If both attributes `#[error(backtrace(forward))]` and `#[error(backtrace)]` are applied to different fields then `f1.backtrace().or(f2)` logic is applied to generated code.  
This way we have all the possible behaviors in a convenient way to use, while not pushing forward harmful anti-patterns by default. Even more, we can explain the reasoning about backtrace forwarding in docs and make an explicit warning about blind applying it everywhere. So, we're not restrict library users in their capabilities, while preferring and pushing forward good practices by default.

**Comment by JelteF** [comment]
Do you want to use the forwarding feature yourself? If so in what case?

Because if not, I'd much rather have only features that are recommended in `derive_more`. As far as I know there's no good reason to use the forwarding feature. That's why I don't want to give people the option to choose something bad. And also I don't want to have code to handle a feature that no-one should use.

**Comment by JelteF** [comment]
> So, we're not restrict library users in their capabilities, while preferring and pushing forward good practices by default.

I think it's good to restrict people in their capabilities of shooting themself and the rust ecosystem in the foot.

**Comment by tyranron** [comment]
@JelteF OK then... I'll try to explain in details our position and errors usage in our codebase.

> The main reason why it's an
anti pattern, is that it causes problems if you want to show the whole
error chain (which is usually what you want). In that case you suddenly
cannot show all the stacktraces in the chain, because there will be
duplicate ones in the chain.

In our codebase (~30K CLOC) we have zero cases of showing the whole error chain separately error-by-error, which is quite illustrative that we do not need it at all. What we do instead, is:
- Every error represents an enum and implements `Display` the way that it renders source error if it's present (with `derive_more` it's really convenient even for generic errors).
- In most situations we forward the innermost `Backtrace` up the error chain. In rare situations we prefer the immediate `Backtrace` instead.
- This way we _separate_ the error displaying and backtracing.
    - To display the whole chain of messages we do not need to dig into `source()`s at all. We just call `Display::fmt` and it conveniently renders all the messages chain down the way. It's a well fit for us in all our use cases (displaying in logs, sending stuff to Sentry, output to a client in a debug mode, etc).
    - We do not display `Backtrace` in `Display::fmt` implementation. We do not need it for regular error messages. But where/when we need it, it's convenient for us to call `backtrace()` at the topmost error and receive the innermost `Backtrace` of the whole chain without digging into `source()`s. The innermost `Backtrace` already contains the whole callstack where errors wrapping happens, so there is no need for us to call `backtrace()` for the all errors in chain, and there is always a single `Backtrace` only for a single error happened.
- If we need to inspect the errors chain, we rarely do it with `source()` methods, but rather go with pattern matching as all our errors are `pub` well-documented enums.

I understand that this way of "doing errors" may be somewhat different from what @mitsuhiko explained, because the problems he has described are non-problems for us. But I cannot call our way of "doing errors" non-idiomatic or bad, just because it plays __too well for us__. And vice versa: I don't see any reason to propose our way of "doing errors" as the most idiomatic and best one, because it plays quite well only for our backend app case, and may be quite inconvenient for library cases or apps of different kinds.

> Do you want to use the forwarding feature yourself? If so in what case?

So, yes. We definitely wanna use forwarding for our codebase. We will use it anyway, but we want `derive_more` to simplify that for us instead of bare hand copy-paste for almost all errors.

And that's why I'm talking about giving that ability for `derive_more` users, even if it's not encouraged as a default way for error handling.

**Comment by JelteF** [comment]
Thanks a lot for the clear explanation. One problem I see with your current approach as is that if you will get a problem if a an Error type from one of your dependencies implements `source`, but not `backtrace` forwarding on an inner error of its own (as is recommended). 
In that case you will not get the `backtrace` by calling `backtrace()` on the outer most error. This is why the community should use the same error handling, so you don't get these problems.

If you would replace your `.backtrace()` calls with calls to the following function, you should get your desired backtrace behaviour (pseudo code, not tested):
```rust
fn first_backtrace(err: Error) -> Option<Backtrace> {
    while err.backtrace() != None {
         if err.source() == None {
             return None
         }
         err = err.source()
    }
    return err.backtrace()
}
```
If you use that you also don't need backtrace forwarding on your own errors. 

So this has the 3 advantages:
1. It works with all dependency errors
2. you don't have to add `#[error(backtrace(forward))]` to your errors (less lines of code)
3. Your errors don't follow an antipattern

Also in your specific case you even have a bonus advantage:
1. You follow the advice of @mitsuhiko, which is one of the main developers of Sentry. So your sentry integration will probably get better automatically.

**Comment by JelteF** [comment]
@tyranron Does my suggestion not fix your usecase for `#[error(backtrace(forward))]` for some reason?

**Comment by JelteF** [comment]
@tyranron @ffuugoo I'm planning to remove the `backtrace(forward)` support from this and merge this in. If you give me a reason why my solution above doesn't work for you then I'm fine with adding it in in a separate PR.

**Comment by tyranron** [comment]
@JelteF sorry for the quite long limbo state of the answer. 😱

First of all:  
I'm totally OK with adding  `backtrace(forward)` support in a separate PR. I like well-separated PRs 🙃

And regarding the question itself:  
I understand the friction around backtrace forwarding for Sentry in the way _they want_ to use it: each error in chain is printed separately, containing its own backtrace and refering to the source. The situation with backtrace forwarding is really nasty for thins. But I'm arguing that the Sentry way of handling this is far from being _the only way_. Even in our case/apps, we're OK and _want to_ use Sentry in slightly different manner, which plays _totally well_ for us (we don't need the all chain, we have that `Displayed` already, we need only the most meaningful `Backtrace`). So having convenient general-purpose library tools is very desirable (at least for us).

Why the alternatives, described above, won't play well for us:
- > If you would replace your `.backtrace()` calls with calls to the following function, you should get your desired backtrace behaviour (pseudo code, not tested):
    
    Yup, but the problem here is that such function is quite often required on _different layers_ of app. So, either these layers should depend on some common `util` (which is bad idea), or some `my_std::ErrorExt` should be introduced in codebase, which handles `.backtrace()` semantics differently. While also, backtrace handling logic may vary depending on the error type itself, so this goes too complex instead of having additional attributes for deriving.

- > One problem I see with your current approach as is that if you will get a problem if a an Error type from one of your dependencies implements source, but not backtrace forwarding on an inner error of its own (as is recommended).
    > In that case you will not get the backtrace by calling backtrace() on the outer most error. This is why the community should use the same error handling, so you don't get these problems.
    
    I don't see this as a problem. The boundary is well controlled in _our_ code and boundary-level errors will treat it correctly. There is no friction with the fact community/library uses another semantics/conventions for error handling. This happens quite often and not only with error handling. You can't eliminate this totally, but it's easy to make adapters between.

Beside the our situation:  
There are situations where backtrace forwarding deriving is still quite helpful. One of examples out of my head is transparent error wrappers, like the one:
```rust
enum DbError {
    Postgres(postgres::Error),
    MySql(mysql::Error),
}
```
Here we _do not want_ the `DbError` error have its own `Display` message, or even represent a some kind of self-contained error, but rather be just a transparent enumeration wrapper for the inner errors. And having `backtrace(forward)` here helps to not write additional code manually, while _explicitly states_ how the error acts with its properties (which _is_ good).

That's why:  
I'm very motivated to have `backtrace(forward)` supported by `derive_more` out-of-the-box. This makes the library _flexible_ for user needs, so convenient to use. Even more, I think it's a good idea to have `backtrace(with = "func_name")` where the custom logic may be handled (like `serde` has for serializing/deserializing) without manual trait implementation.  
I'm also think that this question should be explicitly and well overviewed in documentation.

**Comment by tyranron** [comment]
@ffuugoo please, polish the PR and prepare to merge it without backtrace forwarding.

**Comment by JelteF** [comment]
Thanks for the response @tyranron (no worries on it taking a bit long). Looking forward to the updated PR without the `#[error(backtrace(forward)]`

> There are situations where backtrace forwarding deriving is still quite helpful. One of examples out of my head is transparent error wrappers

I think this would actually be a separate case where you would want to forward all Error methods with something like `#[error(forward)]`. 

I think the reasons you describe why my solution doesn't work make quite a lot of sense and your usecase seems quite common. As far as I can tell the only real solution would be to have the `first_backtrace` function be a method on the `Error` trait with a default implementation. Since it's such a simple function I think it makes sense to try to get that to happen. Could you make a PR to rustc adding that method? (please link the PR here as well) I think having this functionality in the Error trait itself makes more sense than "hacking" around it in this derive crate.

If that PR is not accepted I indeed think it makes sense to add this functionality in this crate and I would accept a PR for for `backtrace(forward)`. In that case we would definitely need quite a bit of explanation in the docs of when and when not to use it. Especially that `backtrace(forward)` isn't suitable for libraries.

**Comment by ffuugoo** [comment]
@JelteF @tyranron Removed backtrace forwarding support. Messed up with force pushing a bit first time. Repushed properly now.

**Comment by tyranron** [comment]
@ffuugoo please, consider CI complains.

**Comment by ffuugoo** [comment]
> @ffuugoo please, consider CI complains.

@JelteF @tyranron Done.

File: src/utils.rs [hunk 2]
```diff
@@ -381,10 +381,16 @@ impl<'input> State<'input> {
...
+        let default_enabled = if trait_name == "Error" {
+            true
+        } else {
+            first_match.map_or(true, |info| !info.enabled.unwrap())
+        };
+
         let defaults = struct_meta_info.to_full(FullMetaInfo {
             // Default to enabled true, except when first attribute has explicit
             // enabling
-            enabled: first_match.map_or(true, |info| !info.enabled.unwrap()),
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/utils.rs]: Explained [here](https://github.com/JelteF/derive_more/issues/92#issuecomment-556081021).

File: src/error.rs [hunk 3]
```diff
@@ -232,43 +322,82 @@ where
...
+    Ok(parsed_fields)
+}
+
+fn parse_field_impl<'a, P, V>(
+    is_valid_default_field_for_attr: &P,
+    len: usize,
+    iter: impl Iterator<Item = (usize, &'a syn::Field, &'a MetaInfo)> + Clone,
+    attr: &str,
+    value: V,
+    out: &mut Option<(usize, &'a syn::Field, &'a MetaInfo)>,
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/error.rs]: I have no idea. Gonna fix it. ¯\_(ツ)_/¯

File: src/utils.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/utils.rs]: Moved comment and add extended explanation as you suggested.

File: src/error.rs [hunk 4]
```diff
@@ -198,31 +230,89 @@ fn parse_fields<'input, 'state>(
...
+                type_params,
+                &state.fields[source].ty,
+            );
+        }
+
+        parsed_fields
+    })
+}
+
+fn is_type_ends_with(ty: &syn::Type, tail: &str) -> bool {
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/error.rs]: Done.

File: src/error.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/error.rs]: Fixed this.

File: src/error.rs [hunk 5]
```diff
@@ -198,31 +230,89 @@ fn parse_fields<'input, 'state>(
...
 
-                ident == "source"
+        DeriveType::Unnamed => {
+            let result = parse_fields_impl(state, |attr, field, len| match attr {
+                "source" => len == 1 && !is_type_ends_with(&field.ty, "Backtrace"),
+                "backtrace" => is_type_ends_with(&field.ty, "Backtrace"),
+                _ => unreachable!(),
+            });
+
+            result.map(|mut parsed_fields| {
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/error.rs]: Fixed.

File: src/error.rs [hunk 6]
```diff
@@ -198,31 +230,89 @@ fn parse_fields<'input, 'state>(
...
+                let len_source_backtrace = (
+                    state.fields.len(),
+                    parsed_fields.source,
+                    parsed_fields.backtrace,
+                );
+
+                if let (2, None, Some(backtrace)) = len_source_backtrace {
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply src/error.rs]: Sort-of fixed it.

Early returns seems more awkward and not much more clear, IMO. So I've moved the code _as it is_ to a separate function and added some comments explaining what's going on.

Ping me if you still prefer early returns, I'll fix it.

File: src/error.rs [hunk 7]
```diff
@@ -289,22 +284,45 @@ fn parse_fields<'input, 'state>(
...
+    // If we have exactly two fields, one of the fields was specified/inferred as backtrace field,
+    // but no source field was specified/inferred...
+    if let (2, None, Some(backtrace)) = len_source_backtrace {
+        // ...then infer other field as source field...
+        let source = (backtrace + 1) % 2;
+        // ...unless it was explicitly marked as non-source.
+        if parsed_fields.data.infos[source].info.source != Some(false) {
+            return Some(source);
+        }
+    }
```
Comment (JelteF) [inline src/error.rs]:
```suggestion
    // If we have exactly two fields...
    if state.fields.len() != 2 {
        return None;
    }
    // ... and no source field was specified/inferred...
    if parsed_fields.source.is_some() {
        return None;
    }

    // ... and one of the fields was specified/inferred as backtrace field,
    if let Some(backtrace) = parsed_fields.backtrace {
        // ...then infer other field as source field...
        let source = (backtrace + 1) % 2;
        // ...unless it was explicitly marked as non-source.
        if parsed_fields.data.infos[source].info.source != Some(false) {
            return Some(source);
        }
    }
```

This was sort of what I meant, but I agree that indeed it doesn't make it much more readable. The comments were most important.

Comment (ffuugoo) [reply src/error.rs]: @JelteF Fixed.

**Comment by JelteF** [comment]
This is now released in version 0.99.5

**Pull request merged**